mlflow/types/schema.py

```diff
@@ -512,9 +513,103 @@ def _merge(self, arr: "Array") -> "Array":
                 f"Array types are incompatible for {self} with dtype={self.dtype} and "
                 f"{arr} with dtype={arr.dtype}"
             )
-        if isinstance(self.dtype, (Array, Object)):
+
+        if (
+            isinstance(self.dtype, (Array, Object, Map))
+            and self.dtype.__class__ is arr.dtype.__class__
+        ):
             return Array(dtype=self.dtype._merge(arr.dtype))
 
+        raise MlflowException(f"Array type {self!r} and {arr!r} are incompatible")
```

Comment:
Do we need this exception?

Reply:
I think so. We should raise proper exception when 2 types can't be merged

Reply:
Ah yes after adding that condition we can reach here :)